particle-iot/device-os
# [SoM] Runtime power management IC detection

### Problem

SoMs may function with or without PMIC / FuelGauge. Currently power management subsystem is enabled only in compile time and is disabled for SoM platforms completely.

### Solution

This PR adds support for runtime presence detection of PMIC and FuelGauge enabling system power management.

This functionality is enabled via a persistent system flag, and by default will be enabled in `tinker` shipped with manufactured devices, making this feature opt-out.

**NOTE: Current development boards for SoMs do not have PMIC interrupt line connected to the MCU, which prevents power management subsystem from functioning correctly**

This PR also fixes a race condition that causes system power manager to re-enable charging every second instead of every 60 seconds.


### Steps to Test

Use the example app below with enabled and disabled `SYSTEM_FLAG_PM_DETECTION`. Observe the log on `Serial1`.

### Example App

```cpp
STARTUP(System.enable(SYSTEM_FLAG_PM_DETECTION));
// STARTUP(System.disable(SYSTEM_FLAG_PM_DETECTION));

Serial1LogHandler dbg(115200, LOG_LEVEL_NONE, {
    {"sys.power", LOG_LEVEL_ALL}
});
```

### References

N/A

---

### Completeness


---

- [Feature] [SoM] Runtime power management IC detection [#1733](https://github.com/particle-iot/device-os/pull/1733)
